### DISCUSSION TITLE
Support for querying props with valid nullability variation

### DISCUSSION THREAD

**Discussion by npbenjohnson** [AutoMapper/AutoMapper.Extensions.OData#136] [Q&A] [answered]
Example scenario:
The entity has a nullable Price field, but the model being exposed is always filtered to have a non-null price. This works fine for mapping and projecting the entity, but when expression mapping is used in the OData library, it breaks because the Price properties are not the same type.
``` csharp
record class Product (int Id, int? Price);
record class PricedProductModel (int Id, int Price);
CreateMap<Product, PricedProductModel>();
``` 
These work:
``` csharp
IMapper.Map<Product, PricedProductModel>();
DbContext.Set<Product>().ProjectTo<PricedProductModel>(...);
``` 

This doesn't, and the error is a little confusing (but I understand why it exists):
``` csharp
DbContext.GetQuery<Product>().GetQueryAsync<PricedProductModel>(...);
``` 
Gives an error that types must match, but the message is a little confusing because the source and destination are the reverse of the only developer-declared CreateMap call:
```
Source: PricedProductModel
Destination: ProductModel
Property: Price
```

I think this is a known unsupported scenario, so I didn't push it up as an issue, but I'm wondering if anyone has a workaround like manually setting up another expression to get used in this scenario, or some info on which library a PR would belong in.

Right now I was trying to hack some expression conversion into the ForAllPropertyMaps on an app because it's migrating from v9 to v11 and a lot of the OData queries stopped working, but it is getting generic-call reflection heavy so I'd rather just find a real solution or document a workaround if a PR would be accepted.

**Comment by BlaiseD** [comment, marked as answer]
You'll need custom expressions when the types are not an exact match e.g.

```c#
		CreateMap<Product, PricedProductModel>()
			.ForMember(dest => dest.Price, opts => opts.MapFrom(x => x.Price.Value));
```

or for constructor maps:

```c#
		CreateMap<Product, PricedProductModel>()
			.ForCtorParam("Price", opts => opts.MapFrom(x => x.Price.Value));
```

If custom expressions don't work then open an issue with sample reproducing the error.

**Reply by BlaiseD** [reply]
I don't think I completely answered your question.  The appropriate library to change the behavior is the [expression mapping](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping) library.  Currently we throw an exception for [mismatches here](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/blob/f86b6e28741064af6fc125c5662559052f4634a7/src/AutoMapper.Extensions.ExpressionMapping/XpressionMapperVisitor.cs#L716).  I think the conversion would be applied further up in [`VisitMember`](https://github.com/AutoMapper/AutoMapper.Extensions.ExpressionMapping/blob/f86b6e28741064af6fc125c5662559052f4634a7/src/AutoMapper.Extensions.ExpressionMapping/XpressionMapperVisitor.cs#L52)

**Reply by npbenjohnson** [reply]
I think casting will work for now, I will look into a fix for the guaranteed valid non-null to nullable cast if you'd consider a PR. 

I've hit a bigger issue with odata where ignoring members breaks querying even if the ignored members aren't referenced in the query. I think the error makes sense in that case too but it would be nice to find a workaround. I'll put up more details in a different discussion next week when I'm back on the project.
